include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -1470,8 +1470,10 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
     void SetBlocked(std::function<bool()> should_block,
                     std::function<void()> post_block) {
       std::unique_lock<std::mutex> lk(mutex);
-      assert(GetStatus() == ThreadStatus::Spinning);
```

Comment:
Since the status can be changed without acquiring the mutex, the assert here does not provide enough guarantee. We need a compare and swap op instead.